src/store_client.cc

```diff
@@ -790,8 +779,13 @@ StoreEntry::invokeHandlers()
 
         CodeContext::Reset(sc->_callback.codeContext);
         debugs(90, 3, "checking client #" << i);
-        storeClientCopy2(this, sc);
+        storeClientCopy2(this, sc.get()); // may invalidate sc CbcPointer
```

Comment:
This is CBDATA where invalidate != free. The std::list contains Pointer explicitly to prevent the problem you refer to.  The list and iterators remain unchanged whether the store_client is free'd and/or just Pointer reset. Only when the loop completes the list gets pruned safely using remove_if.

Reply:
> The list and iterators remain unchanged whether the store_client is free'd and/or just Pointer reset.

If the "may invalidate" comment added by this PR is actually correct, then the above statement is already wrong today. When a store client is destroyed (i.e. when the `sc` pointer is invalidated), Squid calls storeUnregister() which removes the client from the list, invalidating the current loop iterator.

And even if the new "may invalidate" comment is lying about the current code, my change request is still valid because storeClientCopy2() does not _guarantee_ to preserve the store client: Even if that store_client object is preserved today, it may disappear tomorrow in the same call sequence.

> This is CBDATA where invalidate != free

True but irrelevant. It is the membership in the `clients` list that is relevant to my change request, and that membership is linked to `sc` invalidation (i.e. storeUnregister() and its `delete sc`) rather than raw memory deallocation.

Reply:
> > This is CBDATA where invalidate != free
> 
> True but irrelevant.

It is directly relevant to the statement underlying your argument:

> Modifying the container (or, to be more precise, removing the current container item)

There is *no* remove happening to the container itself. Ergo, no UB issue. All that is happening is a flag in the indirectly pointed to data is being modified from true to false so Pointer::get() starts returning nullptr. All container and iterator state remains valid as-was:

initial state:
  list={A, B, C} ; B.raw() == 0x1  ;  B.get() == value.

code does "delete B" to invalidate.

result state:
  list={A, B, C} ;  B.raw() == 0x1  ;  B.get() == nullptr.

Reply:
> There is no remove happening to the container itself

AFAICT, there is or there can be. See the first two paragraphs in my earlier [comment](https://github.com/squid-cache/squid/pull/730#discussion_r492260166) for specifics of how this code gets to storeUnregister() which removes the current item from the container. Nothing in your response addressed those concerns because the response focuses on smart pointer mechanics while the important issue here is storeUnregister().

> code does "delete B" to invalidate.

At the risk of sustaining an irrelevant part of the discussion, I will note that the code does not and cannot `delete B`. `B` (i.e. the thing `sc` in the for-range loop refers to) is a not a raw pointer so it cannot be deleted. `B` is smart pointer object that, in this case, is a part/component of an std::list item object. Nobody deletes `B` because nobody allocates `B` (as a stand-alone object).

What the problematic code actually does (or can do) is call storeClientCopy2(). storeClientCopy2() gives response information to the client-Squid transaction code. Doing so may lead to transaction termination. Transaction termination calls storeUnregister(). storeUnregister() removes `B` from the list (and deletes the corresponding store_client object). All that may happen synchronously, before returning from the storeClientCopy2() call inside the for-range loop iteration, invalidating the current `clients` iterator.